gemfire-core/src/main/java/com/gemstone/gemfire/internal/cache/DiskBlockSortManager.java

```diff
@@ -0,0 +1,260 @@
+        } else {
+          // wait for given timeout
+          if (maxWaitMillis < 0) {
+            maxWaitMillis = Long.MAX_VALUE;
+          }
+          regionStopper.checkCancelInProgress(null);
+          diskStoreStopper.checkCancelInProgress(null);
+          // loop wait to check for cancellation and ignore spurious wakeups
+          if (maxWaitMillis > 0 && this.refCount > 0) {
+            final long start = System.currentTimeMillis();
+            final long loopMillis = Math.min(maxWaitMillis, 10);
+            do {
+              try {
+                this.lock.wait(loopMillis);
+              } catch (InterruptedException ie) {
+                regionStopper.checkCancelInProgress(ie);
+                diskStoreStopper.checkCancelInProgress(ie);
+              }
+            } while (this.refCount > 0 &&
+                (System.currentTimeMillis() - start) < maxWaitMillis);
+          }
+        }
+        this.readOnly = true;
+
+        // capture sorter output once
+        if (this.sorter != null) {
+          Enumerator enumerator = this.sorter.enumerator();
+          Object next;
+          while ((next = enumerator.nextElement()) != null) {
+            this.sorted.add(next);
+          }
+          this.sorter = null;
+        }
+        return new ReaderIdEnumerator(readerId, faultIn);
+      }
+    }
+
+    @GuardedBy("lock")
+    void addNewEntry(DistributedRegion.DiskEntryPage entry) {
+      this.sorter.add(entry);
+    }
+
+    public final class ReaderIdEnumerator
+        implements Enumerator<DistributedRegion.DiskEntryPage> {
+
+      private final int requiredId;
+      private final Iterator<Object> sortedIterator;
+      private final boolean faultIn;
+      private boolean recoveryStopped;
+
+      ReaderIdEnumerator(int readerId, boolean faultIn) {
+        this.requiredId = readerId;
+        this.sortedIterator = sorted.iterator();
+        // no need to check for recovery condition if this is not a full scan
+        this.faultIn = faultIn;
+        this.recoveryStopped = !fullScan && faultIn;
+      }
+
+      @Override
+      public DistributedRegion.DiskEntryPage nextElement() {
+        StoreCallbacks callbacks = CallbackFactoryProvider.getStoreCallbacks();
+        while (this.sortedIterator.hasNext()) {
+          DistributedRegion.DiskEntryPage entry =
+              (DistributedRegion.DiskEntryPage)this.sortedIterator.next();
+          boolean faultInEntry = faultIn;
+          // if this is a full-scan and there is memory available then fault-in
+          // the value in any case to do ordered disk reads as much as possible;
+          // also once recovery is stopped, it will never be checked again
+          if (!recoveryStopped) {
+            recoveryStopped = callbacks.shouldStopRecovery();
+            if (!recoveryStopped) {
+              if (fullScan) {
+                diskStoreStopper.checkCancelInProgress(null);
+                OffHeapHelper.release(entry.readEntryValue());
```

Comment:
Please explain why this OffHeapHelper is coming in picture here? I thought gemxd side OffHeap has nothing to do with snappy/spark way of handling Offheap and this special way of sorting will be just for Snappy side as in the DiskRegionIterator. Gemxd side will continue with DiskSavvyIterator.

Reply:
Going forward the DiskSavyIterator should be removed and replaced with this once this is decently tested. For example GII iteration should also use this because those directly effect SnappyData too.